devito/passes/clusters/blocking.py

```diff
@@ -157,3 +160,102 @@ def decompose(ispace, d, block_dims):
     directions.update({bd: ispace.directions[d] for bd in block_dims})
 
     return IterationSpace(intervals, sub_iterators, directions)
+
+
+class Skewing(Queue):
+
+    def __init__(self, options):
+
+        super(Skewing, self).__init__()
+
+    @timed_pass(name='skewing')
+    def process(self, clusters):
+        return super(Skewing, self).process(clusters)
+
+    def _process_fdta(self, clusters, level, prefix=None):
+
+        return super(Skewing, self)._process_fdta(clusters, level, prefix)
+
+    def callback(self, clusters, prefix):
+        if not prefix:
+            return clusters
+
+        d = prefix[-1].dim
+
+        processed = []
+        for c in clusters:
+            if PARALLEL in c.properties[d] and not d.symbolic_incr.is_Symbol:
+                ispace, exprs = skew(c.ispace, c.exprs, c.properties, d)
```

Comment:
see my comments below first then come back here 


---


In light of what I wrote below, I think we don't need an extra `skew` function. Everything could be written directly here, thus making the implementation much simpler. Instead of traversing the **whole** iteration space (thus having to keep the `skew_pos` around) you could instead traverse the `prefix`, which terminates with `d`, so all Intervals in the prefix are definitely before `d`. Then you don't have to search for anything else... it's just  `mapper[d] = ...`